modules/core/src/core/lifecycle/create-props.js

```diff
@@ -106,25 +112,92 @@ function buildPropDefs(layerClass) {
   return {propTypes, defaultProps};
 }
 
-function buildDefaultProps(props, parentProps, propTypes, layerClass) {
+// Builds a pre-merged default props object that layer props can inherit from
+function buildDefaultPropsObject(props, parentProps, propTypes, layerClass) {
   const defaultProps = Object.create(null);
 
   Object.assign(defaultProps, parentProps, props);
 
-  const descriptors = {};
-
+  // Avoid freezing `id` prop
   const id = getLayerName(layerClass);
   delete props.id;
 
-  Object.assign(descriptors, {
+  // Add getters/setters for async prop properties
+  Object.defineProperties(defaultProps, {
     id: {
       configurable: false,
       writable: true,
       value: id
     }
   });
 
-  Object.defineProperties(defaultProps, descriptors);
+  // Add getters/setters for async prop properties
+  addAsyncPropDescriptors(defaultProps, propTypes);
 
   return defaultProps;
 }
+
+// Create descriptors for overridable props
+function addAsyncPropDescriptors(defaultProps, propTypes) {
+  const defaultValues = {};
+
+  const descriptors = {
+    // Default "resolved" values for async props, returned if value not yet resolved/set.
+    _asyncPropDefaultValues: {
+      enumerable: false,
+      value: defaultValues
+    },
+    // TODO - Shadowed object, just to allow indexing
+    _asyncPropUrls: {
+      enumerable: false,
+      value: {}
+    }
+  };
+
+  // Move async props into shadow values
+  for (const propType of Object.values(propTypes)) {
+    const {name, value} = propType;
+    if (propType.async) {
+      defaultValues[propType.name] = value;
+      Object.assign(descriptors, {
+        [name]: getDescriptorForAsyncProp(name, value)
```

Comment:
Why not just `descriptors[name] = ...`?

Reply:
It was an attempt at breaking the code into short and and well named functions to help the reader. Initially it was one function, but I felt it was long and hard to read.

It is mostly a feeling that this entire file is more complex and subtle than I would like, and since I haven't been able to address that, perhaps I am over indexing on trivial simplifications.